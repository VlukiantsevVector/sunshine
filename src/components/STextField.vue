<template>
  <s-base-input
    :inactive="inactive"
    :hasFocus="hasFocus"
    :isEmpty="isEmpty"
    v-bind="$attrs"
  >

    <input
      class="s-input__input"
      :type="type"
      :value="value"
      :disabled="inactive"
      @input="onInput"
      @keypress="onKeyPress"
      @focus="hasFocus = true"
      @blur="hasFocus = false"
    />

    <!-- expected format -->
    <span
      v-if="hasFocus && format"
      class="s-input__format"
    >
      <span class="s-input__value">{{ value }}</span><span class="s-input__format__remaining">{{ remainingFormat }}</span>
    </span>

  </s-base-input>
</template>

<script lang="ts">
import Vue from 'vue';
import SBaseInput from './SBaseInput.vue';

export default Vue.extend({
  name: 's-text-field',

  components: {
    SBaseInput,
  },

  props: {
    value: {
      type: String,
      default: '',
    },
    phone: {
      type: Boolean,
      default: false,
    },
    password: {
      type: Boolean,
      default: false,
    },
    email: {
      type: Boolean,
      default: false,
    },
    format: {
      type: String,
      default: null,
    },
    inactive: {
      type: Boolean,
      default: false,
    },
  },

  data() {
    return {
      hasFocus: false,
    };
  },

  computed: {
    isEmpty(): boolean {
      return !this.value;
    },

    type(): string {
      switch (true) {
        case this.phone:
          return 'tel';
        case this.password:
          return 'password';
        case this.email:
          return 'email';
        default:
          return 'text';
      }
    },

    remainingFormat(): string {
      if (!this.format) return '';
      return this.format.substring(this.value.length);
    },
  },

  methods: {
    onInput(event: InputEvent) {
      let input = event.target as HTMLInputElement;
      this.$emit('input', input.value);
    },

    onKeyPress(event: KeyboardEvent) {
      if (!this.isValidKey(event.keyCode)) {
        event.preventDefault();
      }
    },

    isValidKey(keyCode: number): boolean {
      if (this.phone) {
        return (
          keyCode === 43 ||
          keyCode === 45 ||
          keyCode >= 48 && keyCode <= 57
        );
      } else {
        return true;
      }
    },
  },
});
</script>
