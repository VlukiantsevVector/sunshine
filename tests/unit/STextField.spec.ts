import { mount } from '@vue/test-utils';
import STextField from '@/components/STextField.vue';

describe('STextField', () => {
  it('renders empty text field', () => {
    const wrapper = mount(STextField, {
      propsData: {
        label: 'Text field',
        value: '',
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders text field with text', () => {
    const wrapper = mount(STextField, {
      propsData: {
        label: 'Text field',
        value: 'foobar',
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('emits entered value', () => {
    const wrapper = mount(STextField, {
      propsData: {
        label: 'Text field',
        value: '',
      },
    });
    let input = wrapper.findAll('input');
    input.setValue('kokobaba');
    input.trigger('change');

    expect(wrapper.emitted('input')[0]).toEqual(['kokobaba']);
  });
});
