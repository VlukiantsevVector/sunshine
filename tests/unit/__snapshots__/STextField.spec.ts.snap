// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`STextField renders text field 1`] = `
<span class="s-input" label="Text field"><span class="s-input__label s-input__label--aside">Text field</span> <input type="text" class="s-input__input">
<!----></span>
`;
